src/clj/rems/poller/email.clj

```diff
@@ -0,0 +1,219 @@
+                 :application/request-id "r1"
+                 :application/commenters ["commenter1" "commenter2"]}
+                {:application/id 7
+                 :event/type :application.event/member-joined
+                 :event/actor "somebody"}
+                {:application/id 7
+                 :event/type :application.event/commented
+                 :event/actor "commenter2"
+                 :application/request-id "r1"
+                 :application/comment ["this is a comment"]}
+                {:application/id 7
+                 :event/type :application.event/member-added
+                 :event/actor "handler"
+                 :application/member {:userid "member"}}
+                {:application/id 7
+                 :event/type :application.event/decision-requested
+                 :event/actor "assistant"
+                 :application/request-id "r2"
+                 :application/deciders ["decider"]}
+                {:application/id 7
+                 :event/type :application.event/decided
+                 :event/actor "decider"
+                 :application/decision :approved}
+                {:application/id 7
+                 :event/type :application.event/approved
+                 :event/actor "handler"}]
+        application (dynamic/apply-events nil events)]
+    (is (= [[] ;; created
+            [] ;; submitted
+            [{:to "somebody@example.com" :body "invitation email"}]
+            [{:to "commenter1" :body "please comment on 7"}
+             {:to "commenter2" :body "please comment on 7"}]
+            [] ;; member-joined
+            [{:to "handler"
+              :body "comment by commenter2: [\"this is a comment\"]"}
+             {:to "assistant"
+              :body "comment by commenter2: [\"this is a comment\"]"}]
+            [{:to "member" :body "you've been added"}]
+            [{:to "decider" :body "please decide 7"}]
+            [{:to "handler" :body "decision by decider: :approved"}
+             {:to "assistant" :body "decision by decider: :approved"}]
+            [{:to "applicant" :body "application 7 has been approved"}
+             {:to "somebody", :body "application 7 has been approved"}
+             {:to "member", :body "application 7 has been approved"}]]
+           (mapv #(event-to-emails-impl % application) events)))))
+
+;;; Sending emails
+
+(defn send-email! [email-spec]
+  ;; just a stub for now
+  (log/info "email:" (pr-str email-spec)))
+
+;;; Generic poller infrastructure
+
+;; these can be moved to rems.poller once we have multiple pollers
+(defn get-poller-state [name-kw]
+  (or (json/parse-string (:state (db/get-poller-state {:name (name name-kw)})))
+      {:event/id -1}))
+
+(defn set-poller-state! [name-kw state]
+  (db/set-poller-state! {:name (name name-kw) :state (json/generate-string state)})
+  nil)
+
+(defn run-event-poller [name-kw process-event!]
+  (let [prev-state (get-poller-state name-kw)
```

Comment:
This will work only when there is just one application instance running. Is that a realistic assumption? To support multiple instances for redundancy, one solution would be to wrap the poller a transaction and use "select for update".

Reply:
ScheduledThreadPoolExecutor guarantees exclusion, I'll add a comment

Reply:
But not when there are multiple JVMs/servers.